framework/src/ble_manager/ble_manager_state.c

```diff
@@ -197,6 +267,137 @@ ble_result_e blemgr_handle_request(blemgr_msg_s *msg)
 		ret = ble_drv_conn_is_any_active(is_active);
 	} break;
 
+	// Scanner
+	case BLE_EVT_CMD_START_SCAN: {
+		BLE_STATE_CHECK;
+
+		if (g_scan_ctx.state != BLE_SCAN_STOPPED) {
+			ret = TRBLE_BUSY;
+			break;
+		}
+
+		ble_scan_state_e priv_state = g_scan_ctx.state;
+		blemgr_msg_params *param = (blemgr_msg_params *)msg->param;
+		ble_scan_filter *filter = (ble_scan_filter *)param->param[0];
+		ble_scan_callback_list *callbacks = (ble_scan_callback_list *)param->param[1];
+
+		if (filter != NULL) {
+			memcpy(&(g_scan_ctx.filter), filter, sizeof(ble_scan_filter));
+		}
+
+		if (callbacks == NULL) {
+			ret = TRBLE_INVALID_ARGS;
+			break;
+		}
+		memcpy(&(g_scan_ctx.callback), callbacks, sizeof(ble_scan_callback_list));
+
+		g_scan_ctx.state = BLE_SCAN_CHANGING;
+		ret = ble_drv_start_scan((trble_scan_filter *)filter);
+		if (ret != TRBLE_SUCCESS) {
+			memset(&(g_scan_ctx.callback), 0, sizeof(ble_scan_callback_list));
+			memset(&(g_scan_ctx.filter), 0, sizeof(ble_scan_filter));
+			g_scan_ctx.state = priv_state;
+		}
+	} break;
+
+	case BLE_EVT_CMD_STOP_SCAN: {
+		BLE_STATE_CHECK;
+
+		ret = TRBLE_SUCCESS;
+		if (g_scan_ctx.state != BLE_SCAN_STARTED) {
+			break;
+		}
+
+		ble_scan_state_e priv_state = g_scan_ctx.state;
+
+		g_scan_ctx.state = BLE_SCAN_CHANGING;
+		ret = ble_drv_stop_scan();
+		if (ret != TRBLE_SUCCESS) {
+			g_scan_ctx.state = priv_state;
+		}
+	} break;
+
+	case BLE_EVT_CMD_WHITELIST_ADD: {
+		BLE_STATE_CHECK;
+
+		ret = TRBLE_SUCCESS;
+		trble_addr *addr = (trble_addr *)msg->param;
+		if (addr == NULL) {
+			ret = TRBLE_INVALID_ARGS;
+			break;
+		}
+
+		if (_whitelist_is_exist(addr) == true) {
+			/* Already Exist */
+			break;
+		}
+
+		if (_whitelist_is_full() == true) {
+			ret = TRBLE_OUT_OF_MEMORY;
+			break;
+		}
+
+		if (g_scan_ctx.state != BLE_SCAN_STOPPED) {
```

Comment:
How about moving this condition check at the top of the case statement like line 322?
Then we can save the time, if the state is not `BLE_SCAN_STOPPED`.

Reply:
By checking whitelist first, an user can get a exact result(already registered, whitelist full) regardless of state